### DISCUSSION TITLE
Unfortunate non-mnemonic for ranges?

### DISCUSSION THREAD

**Discussion by gwhitney** [DanielXMoore/Civet#999] [General]
I always get `0..n` and `0...n` backwards because the one with more dots looks like it should include more, so I assume that's the one that includes the right endpoint and the one with fewer dots is the one that leaves it out. But actually it's the other way 'round. I doubt at this point there's much to be done about that, except I am hoping someone here might have another mnemonic to suggest to displace this "obvious" one from my head. Thoughts?

**Comment by bbrk24** [comment]
I think we have to keep these two, if nothing else for coffeeCompat mode. However, we could alias one or both of them:
- Rust uses `..=` for the inclusive one
- Swift uses `..<` for the exclusive one

**Comment by STRd6** [comment]
Civet inherited ranges from CoffeeScript which I believe got them from Ruby which got them from Perl which I think got them from AWK or maybe Bash sequence expansions.

Currently Civet, CoffeeScript, and Ruby have the same/similar range semantics.

I think Ruby semantics being well established is a good enough reason to keep it as is unless there is a much stronger and well known consensus in other popular languages.

Adding aliases could be a reasonable compromise as well.

**Comment by edemaine** [comment]
FWIW, `a..b` inclusive goes back to at least Pascal, which is one of the first languages I learned. So that's how I remember. :slightly_smiling_face:

My best attempt at a mnemonic:
* `..` is pronounced "up to"
* `...` is pronounced "up to but not including"

Personally, I mainly use `...`, as Python taught me well to think exclusive on the end. If you use one mainly, it's easier to remember what the other does.

We've discussed adding Python's `:` as an alternative to `...`, which might help if you come from Python. But it conflicts with object notation.

**Reply by bbrk24** [reply]
> * `..` is pronounced "up to"
> 
> * `...` is pronounced "up to but not including"

Swift calls these ["Through"](https://developer.apple.com/documentation/swift/partialrangethrough) and ["UpTo"](https://developer.apple.com/documentation/swift/partialrangeupto) respectively, at least for the ranges with no lower bound. (The ranges with an explicit lower bound are more boringly "Range" and "ClosedRange".)

I've worked in [Aussie++](https://github.com/zackradisic/aussieplusplus), which is mostly a joke language, but it had a decent syntax for ranges: `[x TO y]` and `[x TO y)`. The keyword `TO` (traditionally but not necessarily all-caps, like in SQL) is what makes this a range and allows the mismatched brackets. I believe you can even say `(x TO y)` and `(x TO y]`, but these aren't needed as much in practical code. This comes directly from mathematical interval notation so it likely has a bit older precedent than Pascal -- though they're typically spelled with a comma rather than the word "TO", that would probably make the grammar ambiguous.

**Reply by edemaine** [reply]
We could consider `[x..y)` (and of course the two other combinations). I don't think this is ambiguous, and it naturally extends `[x..y]`.

(I know we've discussed mismatched brackets before, but I don't think we've discussed it with `..` before...)

**Reply by gwhitney** [reply]
Well for a mathematician `[..]` and `[..)` are certainly very comfortable, so I would be a happy camper if that were implemented. :-) And I would be perfectly content with `(..]` as well, as for example I am working on code right now where I have to write `[n-1..0]` when `(n..0]` would absolutely make the intent clearer. But I do worry that allowing the ordinary paren to open could make parsing more difficult (?) As for `(..)` I am a little hard pressed to think of a coding use, and it seemed even more worrisome parsing-wise, but if there turn out to be no serious parsing issues, then having all four is nicely symmetric and complete.

**Reply by bbrk24** [reply]
`[x..y]` and `(x..y)` are a little ambiguous because they are, in some cases, actually valid JS:

```js
const toString = 6;
console.log(5..toString); // function toString()
```

**Reply by edemaine** [reply]
Luckily, this is not the case in Civet: `5..toString()` is invalid input. It is *output* for the more natural input `5.toString()`. (I should document this in the Comparison page!)

I don't think `[x..y)` and `(x..y]` are harder to parse, thanks to the wonder of PEG parsing. But I guess I'll find out when trying to implement it. :slightly_smiling_face:

**Reply by vendethiel** [reply]
`[)` and `(]` look very confusing to non-mathy people and simple tools, this'll require special support in most editors. Is a range syntax really worth breaking a lot of (or, perhaps, most) people's expectations?

**Reply by edemaine** [reply]
I agree that it will break bracket matching. Still might be worth it to some... 

Some other options are presented here: https://softwareengineering.stackexchange.com/questions/64975/designing-the-perfect-range-literal

I don't think we can use `!` to exclude an endpoint, but maybe some other notation...

**Reply by vendethiel** [reply]
Raku (ex Perl 6) uses `1 ..^ 6` for exclusive end. (and `1 ^.. 6` for exclusive start, as well as `1 ^..^ 6` if you want to be thorough...)

**Comment by gwhitney** [comment]
In case it helps anyone else, unless/until we get some "nicer" notation (I also like bbrk24's original suggestion of adopting both `[a..<b]` and `[a..=b]` from Rust and Swift, and by symmetry maybe `[a<..b]` and even in case of counting down `[a>..b]` and/or `[a..>b]`), my mantra is

"More dots, fewer elements."

**Comment by edemaine** [comment]
We have a whole new suite of slice notation via #1476!
